### ISSUE TITLE
Invalid data passed as a struct "pair of doubles" using reflection on Linux 1.0.x runtime

### ISSUE THREAD

**Issue by jskeet** [dotnet/coreclr#14391]
Originally reported in this blog post: https://codeblog.jonskeet.uk/2017/10/08/diagnosing-a-linux-only-unit-test-failure/

Basically, I'm calling a static method via reflection, passing in a simple struct that consists of two `double` values. On netcoreapp1.0 *on Linux only*, the value the method receives has a bogus second value.

Project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>netcoreapp1.0;netcoreapp1.1;netcoreapp2.0</TargetFrameworks>
  </PropertyGroup>
</Project>
```

Source code:

```csharp
using System;
using System.Reflection;

class Program
{
   static void Main(string[] args)
   {
       DoublePair pair = new DoublePair(1, 2);
       var method = typeof(DoublePair).GetMethod("Print");
       method.Invoke(null, new object[] { pair });
   }
}

struct DoublePair
{
    private readonly double x;
    private readonly double y;
    
    public DoublePair(double x, double y)
    {
        this.x = x;
        this.y = y;
    }
    
    public static void Print(DoublePair pair) =>
        Console.WriteLine($"{pair.x} {pair.y}");
}
```

Expected output: 1 2

Sample output on netcoreapp1.0 **on Linux**: 1 1.83063495561695E-316

I get the expected output in netcoreapp1.1 and netcoreapp2.0, but not in netcoreapp1.0. Currently using fx 1.0.7 as installed via dotnet-dev-1.1.4. Everything works fine in Windows.

Applying explicit struct layout to `DoublePair` fixes the issue, suggesting an invalid JIT optimization somewhere.

**Comment by eerhardt** [comment]
I discussed this issue with @janvorli this morning and he said the fix was https://github.com/dotnet/coreclr/pull/7716, which was ported to 1.1.x in https://github.com/dotnet/coreclr/pull/7748, but was never ported to 1.0.x.

Here's the original reported issue: https://github.com/dotnet/coreclr/issues/7685.

@Petermarcu - do we want to port this fix to 1.0.x?  It seems like it would cause hard to diagnose problems in user's apps.

**Comment by mihailik** [comment]
Does this affect `KeyValuePair<double,double>`?

I guess less people care for little pesky structs. But keeping `Dictionary<double,double>` out of harm's way is a different business.

**Comment by stephentoub** [comment]
> But keeping Dictionary<double,double> out of harm's way is a different business.

Dictionary doesn't use reflection invocation.

**Comment by mihailik** [comment]
@stephentoub Dictionary doesn't, but people using D<double,double> possibly do — consider seiralization scenarios.

**Comment by mihailik** [comment]
Here's a live environment with exact reproduction (dotnet 1.0, 1.1 and 2.0 installed), on Cloud9:

https://ide.c9.io/mihailik/dotnet-bug

**Comment by jkotas** [comment]
The fix was ported to 1.0.x. It will show up in next 1.0.x servicing release.
